# G2M Insight for Cab Investment Firm

## 1. Project Overview

XYZ is a private firm in US. Due to remarkable growth in the Cab Industry in last few years and multiple key players in the market, it is planning for an investment in Cab industry. <p>

**Objective**: Provide actionable insights to help XYZ firm in identifying the right company for making investment.

## 2. Data Cleaning

Multiple data sets that contain information on 2 cab companies have been provided. Each dataset provided represents different aspects of the customer profile. Time period of data is from 2016-01-31 to 2018-12-31. <p>
    
Below are the list of datasets which are provided for the analysis:

|Dataset|Description|    
|:---|:---|    
|`Cab_Data.csv`|This file includes details of transaction for 2 cab companies|
|`Customer_ID.csv`|This is a mapping table that contains a unique identifier which links the customer’s demographic details|
|`Transaction_ID.csv`|This is a mapping table that contains transaction to customer mapping and payment mode|
|`City.csv`|This file contains list of US cities, their population and number of cab users|

### Import Libraries & Datasets

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import datetime 

In [2]:
cab = pd.read_csv('datasets/Cab_Data.csv')
city = pd.read_csv('datasets/City.csv')
customer_id = pd.read_csv('datasets/Customer_ID.csv')
transaction_id = pd.read_csv('datasets/Transaction_ID.csv')

### Cab Dataset

In [3]:
cab.head()

Transaction ID  Date of Travel   Company        City  KM Travelled  \
0        10000011           42377  Pink Cab  ATLANTA GA         30.45   
1        10000012           42375  Pink Cab  ATLANTA GA         28.62   
2        10000013           42371  Pink Cab  ATLANTA GA          9.04   
3        10000014           42376  Pink Cab  ATLANTA GA         33.17   
4        10000015           42372  Pink Cab  ATLANTA GA          8.73   

   Price Charged  Cost of Trip  
0         370.95       313.635  
1         358.52       334.854  
2         125.20        97.632  
3         377.40       351.602  
4         114.62        97.776

In [4]:
cab.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 359392 entries, 0 to 359391
Data columns (total 7 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Transaction ID  359392 non-null  int64  
 1   Date of Travel  359392 non-null  int64  
 2   Company         359392 non-null  object 
 3   City            359392 non-null  object 
 4   KM Travelled    359392 non-null  float64
 5   Price Charged   359392 non-null  float64
 6   Cost of Trip    359392 non-null  float64
dtypes: float64(3), int64(2), object(2)
memory usage: 19.2+ MB


While there are no missing values in this dataset, we have noticed that the `Date of Travel` column records the number of days instead of the date in the year-month-day format. We need to investigate and correct this issue.

In [5]:
print(cab['Date of Travel'].min())
print(cab['Date of Travel'].max())

42371
43465


Even though no additional information was provided, the earliest data in the `Date of Travel` column appears to be approximately 42000, which is equivalent to roughly 115 years. Assuming the earliest data is from 2016, 115 years before that is 1901. Therefore, we can hypothesize that the `Date of Travel` column is recorded as the number of days that have passed since 1900-01-01. In order to convert this column into the correct date format, we need to apply a transformation to convert the number of days into the year-month-day format.

In [6]:
baseline_date = pd.to_datetime('1900-01-01')
cab['Date of Travel'] = baseline_date + pd.to_timedelta(cab['Date of Travel'], unit='D')

In [7]:
cab.head()

Transaction ID Date of Travel   Company        City  KM Travelled  \
0        10000011     2016-01-10  Pink Cab  ATLANTA GA         30.45   
1        10000012     2016-01-08  Pink Cab  ATLANTA GA         28.62   
2        10000013     2016-01-04  Pink Cab  ATLANTA GA          9.04   
3        10000014     2016-01-09  Pink Cab  ATLANTA GA         33.17   
4        10000015     2016-01-05  Pink Cab  ATLANTA GA          8.73   

   Price Charged  Cost of Trip  
0         370.95       313.635  
1         358.52       334.854  
2         125.20        97.632  
3         377.40       351.602  
4         114.62        97.776

In [8]:
cab.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 359392 entries, 0 to 359391
Data columns (total 7 columns):
 #   Column          Non-Null Count   Dtype         
---  ------          --------------   -----         
 0   Transaction ID  359392 non-null  int64         
 1   Date of Travel  359392 non-null  datetime64[ns]
 2   Company         359392 non-null  object        
 3   City            359392 non-null  object        
 4   KM Travelled    359392 non-null  float64       
 5   Price Charged   359392 non-null  float64       
 6   Cost of Trip    359392 non-null  float64       
dtypes: datetime64[ns](1), float64(3), int64(1), object(2)
memory usage: 19.2+ MB


Lastly, we should remove duplicate if there is any.

In [9]:
cab = cab.drop_duplicates(keep='first')

### City Dataset

In [10]:
city

City   Population      Users
0        NEW YORK NY   8,405,837    302,149 
1         CHICAGO IL   1,955,130    164,468 
2     LOS ANGELES CA   1,595,037    144,132 
3           MIAMI FL   1,339,155     17,675 
4     SILICON VALLEY   1,177,609     27,247 
5      ORANGE COUNTY   1,030,185     12,994 
6       SAN DIEGO CA     959,307     69,995 
7         PHOENIX AZ     943,999      6,133 
8          DALLAS TX     942,908     22,157 
9         ATLANTA GA     814,885     24,701 
10         DENVER CO     754,233     12,421 
11         AUSTIN TX     698,371     14,978 
12        SEATTLE WA     671,238     25,063 
13         TUCSON AZ     631,442      5,712 
14  SAN FRANCISCO CA     629,591    213,609 
15     SACRAMENTO CA     545,776      7,044 
16     PITTSBURGH PA     542,085      3,643 
17     WASHINGTON DC     418,859    127,001 
18      NASHVILLE TN     327,225      9,270 
19         BOSTON MA     248,968     80,021

In [11]:
city.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   City        20 non-null     object
 1   Population  20 non-null     object
 2   Users       20 non-null     object
dtypes: object(3)
memory usage: 608.0+ bytes


We have observed that the `Population` and `Users` columns are recorded as strings. To ensure consistency and enable numerical operations, we need to convert them to integer data type.

In [12]:
city['Population'] = city['Population'].str.replace(',', '')
city['Population'] = city['Population'].astype(int)

city['Users'] = city['Users'].str.replace(',', '')
city['Users'] = city['Users'].astype(int)

In [13]:
city.head()

City  Population   Users
0     NEW YORK NY     8405837  302149
1      CHICAGO IL     1955130  164468
2  LOS ANGELES CA     1595037  144132
3        MIAMI FL     1339155   17675
4  SILICON VALLEY     1177609   27247

In [14]:
city.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   City        20 non-null     object
 1   Population  20 non-null     int64 
 2   Users       20 non-null     int64 
dtypes: int64(2), object(1)
memory usage: 608.0+ bytes


### Customer ID Dataset

In [15]:
customer_id.head()

Customer ID Gender  Age  Income (USD/Month)
0        29290   Male   28               10813
1        27703   Male   27                9237
2        28712   Male   53               11242
3        28020   Male   23               23327
4        27182   Male   33                8536

In [16]:
customer_id.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49171 entries, 0 to 49170
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Customer ID         49171 non-null  int64 
 1   Gender              49171 non-null  object
 2   Age                 49171 non-null  int64 
 3   Income (USD/Month)  49171 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 1.5+ MB


We can see there are no missing values from the dataset. We should remove duplicates if there are any.

In [17]:
customer_id = customer_id.drop_duplicates(keep='first')

### Transaction ID Dataset

In [19]:
transaction_id.head()

Transaction ID  Customer ID Payment_Mode
0        10000011        29290         Card
1        10000012        27703         Card
2        10000013        28712         Cash
3        10000014        28020         Cash
4        10000015        27182         Card

In [20]:
transaction_id.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 440098 entries, 0 to 440097
Data columns (total 3 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   Transaction ID  440098 non-null  int64 
 1   Customer ID     440098 non-null  int64 
 2   Payment_Mode    440098 non-null  object
dtypes: int64(2), object(1)
memory usage: 10.1+ MB


We can see there are no missing values from the dataset. We should remove duplicates if there are any.

In [21]:
transaction_id = transaction_id.drop_duplicates(keep='first')

### Join Tables

Now that we have cleaned and fixed our four tables, we can join them together to facilitate analysis.

In [22]:
df = cab.merge(transaction_id, on = 'Transaction ID').merge(customer_id, on = 'Customer ID').merge(city, on = 'City')
df.head()

Transaction ID Date of Travel     Company        City  KM Travelled  \
0        10000011     2016-01-10    Pink Cab  ATLANTA GA         30.45   
1        10351127     2018-07-23  Yellow Cab  ATLANTA GA         26.19   
2        10412921     2018-11-25  Yellow Cab  ATLANTA GA         42.55   
3        10000012     2016-01-08    Pink Cab  ATLANTA GA         28.62   
4        10320494     2018-04-23  Yellow Cab  ATLANTA GA         36.38   

   Price Charged  Cost of Trip  Customer ID Payment_Mode Gender  Age  \
0         370.95      313.6350        29290         Card   Male   28   
1         598.70      317.4228        29290         Cash   Male   28   
2         792.05      597.4020        29290         Card   Male   28   
3         358.52      334.8540        27703         Card   Male   27   
4         721.10      467.1192        27703         Card   Male   27   

   Income (USD/Month)  Population  Users  
0               10813      814885  24701  
1               10813      814885  24701  
2               10813      814885  24701  
3                9237      814885  24701  
4                9237      814885  24701

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 359392 entries, 0 to 359391
Data columns (total 14 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   Transaction ID      359392 non-null  int64         
 1   Date of Travel      359392 non-null  datetime64[ns]
 2   Company             359392 non-null  object        
 3   City                359392 non-null  object        
 4   KM Travelled        359392 non-null  float64       
 5   Price Charged       359392 non-null  float64       
 6   Cost of Trip        359392 non-null  float64       
 7   Customer ID         359392 non-null  int64         
 8   Payment_Mode        359392 non-null  object        
 9   Gender              359392 non-null  object        
 10  Age                 359392 non-null  int64         
 11  Income (USD/Month)  359392 non-null  int64         
 12  Population          359392 non-null  int64         
 13  Users               359392 no

## 3. Preliminary Insights

### Question 1

